walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -21,69 +21,113 @@
+      } else if (!first.contains(second)) {
+        combined = TimeRange.fromStartEnd(first.start(), second.end(), false);
+        useCombinedRange = true;
+      } else {
+        conflictingTimes.add(first);
+        nested = true;
 
-        conflictingTimes.remove(second);
-        i--;
+        // Skip over adding the second TimeRange
+        i++;
       }
     }
 
-    // Find every TimeRange that doesn't have a conflict
-    if (conflictingTimes.isEmpty()) {
-      possibleTimes.add(TimeRange.WHOLE_DAY);
+    if (conflicts.size() == 0 || nested) {
+      return conflictingTimes;
+    }
+
+    if (conflicts.size() == 1) {
+      conflictingTimes.add(conflicts.get(0));
+    } else if (useCombinedRange) {
+      conflictingTimes.add(combined);
     } else {
-      TimeRange initial = conflictingTimes.get(0);
-      TimeRange valid = TimeRange.fromStartEnd(TimeRange.START_OF_DAY, initial.start(), false);
-      possibleTimes.add(valid);
+      conflictingTimes.add(second);
+    }
 
-      for (int i = 0; i < conflictingTimes.size() - 1; i++) {
-        TimeRange first = conflictingTimes.get(i);
-        TimeRange second = conflictingTimes.get(i + 1);
+    return conflictingTimes;
+  }
 
-        valid = TimeRange.fromStartEnd(first.end(), second.start(), false);
-        possibleTimes.add(valid);
-      }
+  public List<TimeRange> findPossibleTimes(List<TimeRange> conflictingTimes, long meetingDuration) {
+    List<TimeRange> possibleTimes = new ArrayList<>();
 
-      TimeRange last = conflictingTimes.get(conflictingTimes.size() - 1);
-      valid = TimeRange.fromStartEnd(last.end(), TimeRange.END_OF_DAY, true);
-      possibleTimes.add(valid);
+    if (conflictingTimes.isEmpty()) {
+      addTimeIfLongEnough(TimeRange.WHOLE_DAY, meetingDuration, possibleTimes);
+      return possibleTimes;
     }
 
-    // Remove every TimeRange that is too short
-    long duration = request.getDuration();
+    TimeRange initial = conflictingTimes.get(0);
+    TimeRange valid = TimeRange.fromStartEnd(TimeRange.START_OF_DAY, initial.start(), false);
+    addTimeIfLongEnough(valid, meetingDuration, possibleTimes);
 
-    for (int i = 0; i < possibleTimes.size(); i++) {
-      if (possibleTimes.get(i).duration() < duration) {
-        possibleTimes.remove(i);
-        i--;
-      }
+    for (int i = 0; i < conflictingTimes.size() - 1; i++) {
+      TimeRange first = conflictingTimes.get(i);
+      TimeRange second = conflictingTimes.get(i + 1);
+
+      valid = TimeRange.fromStartEnd(first.end(), second.start(), false);
+      addTimeIfLongEnough(valid, meetingDuration, possibleTimes);
     }
 
+    TimeRange last = conflictingTimes.get(conflictingTimes.size() - 1);
+    valid = TimeRange.fromStartEnd(last.end(), TimeRange.END_OF_DAY, true);
+    addTimeIfLongEnough(valid, meetingDuration, possibleTimes);
+
     return possibleTimes;
   }
+
+  public void addTimeIfLongEnough(TimeRange time, long minDuration, List<TimeRange> timeList) {
```

Comment:
I don't recommend passing a list to be mutated. Instead, it's more clear to return the final list because of better encapsulation, thread safety, and avoiding having invalid state.